synapse/cortex.py

```diff
@@ -1252,6 +1286,137 @@ async def initServiceStorage(self):
 
         await self.auth.addAuthGate('cortex', 'cortex')
 
+    async def _storUpdateMacros(self):
+        async for name, node in self.hive.open(('cortex', 'storm', 'macros')):
+            info = (await node.dict()).pack()
+            await self.addStormMacro(info)
+
+    def getStormMacro(self, name):
+
+        byts = self.slab.get(name.encode(), db=self.macrodb)
+        if byts is None:
+            return None
+
+        return s_msgpack.un(byts)
+
+    def reqStormMacro(self, name):
+        mdef = self.getStormMacro(name)
+        if mdef is None:
+            raise s_exc.NoSuchName(mesg=f'Macro name not found: {name}')
+        return mdef
+
+    def _reqStormMacroPerm(self, user, name, level):
+
+        mdef = self.reqStormMacro(name)
+
+        userlevel = mdef['permissions']['users'].get(user.iden)
+        if userlevel is not None and userlevel >= level:
+            return mdef
+
+        roleperms = mdef['permissions']['roles']
+        for role in user.getRoles():
+            rolelevel = roleperms.get(role.iden)
+            if rolelevel is not None and rolelevel >= level:
+                return mdef
+
+        # allow read by default unless explicitly set
+        if level <= PERM_READ:
+            return mdef
```

Comment:
could do this before checking all the perms

Reply:
I would like to have PERM_NONE if explicitly set take precedence to allow denying even read

Reply:
ah okay, so like in the above checks it should be raising AuthDeny if there is a user/role level set and it is `< level`?

Reply:
Yep! Will tweak, thanks.